### PR TITLE
Add integration test of static policy plugin

### LINKED ISSUE #874: As a system administator, I want to verify that the GEOPM static policy plugin applies limits to non-GEOPM jobs
> estimate 5

### Use Case
A system admin or site admin may want to apply limits such as a power cap or maximum frequency to all jobs on the system, whether they use GEOPM or not.  This can be enabled through the use of a resource manager plugin similar to https://github.com/geopm/geopm-slurm/blob/master/src/geopm_spank.c that uses the geopm_agent_enforce_policy() feature of GEOPM.  An administrator may want to test whether the plugin is installed and configured correctly with the agent specified in GEOPM's environment files.  Two common use cases should be covered:
- power limit using either PowerBalancerAgent or PowerGovernorAgent
- maximum or fixed frequency using either FrequencyMapAgent or EnergyEfficientAgent

### Done Criteria
- A test that runs a job script with a high-power operation without any GEOPM controller and reports power cap or frequency cap as detected by reading the corresponding MSRs.  The test should check these values against the expected values using the configuration specified in the environment-[default | override] file.

### PR DESCRIPTION
**Pull request by bakerbrandond**
- Asserts that expected site policy is enforced on non-GEOPM jobs
- Fixes #874

Change-Id: I32b4197323e2814e20ba122ab4f32b6ca2eef604
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

### CODE REVIEW

File: test_integration/test_plugin_static_policy.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+from test_integration import geopm_context
+import geopmpy.io
+from test_integration import geopm_test_launcher
+from test_integration import util
+import geopmpy.topo
+
+# todo this function results in test launcher
+# invocation, and is called 3 times (1 by class and
+# in 2 decorator tests).  may be useful to add once logic
```
Comment (bgeltz) [inline test_integration/test_plugin_static_policy.py]: Is there an issue open for this?

Comment (bgeltz) [inline test_integration/test_plugin_static_policy.py]: Is there an issue open for this open already?

File: test_integration/test_plugin_static_policy.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    """
+    @classmethod
+    def setUpClass(cls):
+        cls._tmp_files = []
+        cls._geopmadminagent = geopmadminagent()
+        cls._geopmadminagentpolicy = geopmadminagentpolicy()
+
+    @unittest.skipUnless(geopmadminagent() in ['frequency_map', 'energy_efficient'],
+                         'Requires environment defaut/override to be configured to cap frequency.')
```
Comment (bgeltz) [inline test_integration/test_plugin_static_policy.py]: Spelling: defau*L*t/override

File: test_integration/test_plugin_static_policy.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        elif self._geopmadminagent == 'energy_efficient':
+            test_freq = self._geopmadminagentpolicy['FREQ_FIXED']
+        current_freq = geopm_test_launcher.geopmread("MSR::PERF_CTL:FREQ board 0")
+        self.assertEqual(test_freq, current_freq)
+
+    @unittest.skipUnless(geopmadminagent() in ['power_governor','power_balancer'],
+                         'Requires environment defaut/override to be configured to cap power.')
+    def test_power_cap_enforced(self):
+        test_power = self._geopmadminagentpolicy['POWER_PACKAGE_LIMIT_TOTAL']
+        # todo need test_launcher to execute topo on compute node
```
Comment (bgeltz) [inline test_integration/test_plugin_static_policy.py]: Issue for this too please.

File: test_integration/test_plugin_static_policy.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        num_pkg = geopmpy.topo.num_domain('package')
+        current_power = num_pkg * geopm_test_launcher.geopmread("MSR::PKG_POWER_LIMIT:PL1_POWER_LIMIT board 0")
```
Comment (bgeltz) [inline test_integration/test_plugin_static_policy.py]: It isn't safe to assume that the PL1 limit is the same on both sockets if multisocket.

I was going to suggest you use num_pkg to loop over the packages, and use geopmread to read the limit on the individual sockets.  But I see you're reading the board limit...

I think this should fail:  reading the board limit should read each package and sum them I think.  There's a bug somewhere here, I think it's in how geopmread sums (or not) the values for the board domain if this logic works.

**Review by bgeltz — CHANGES_REQUESTED** [review]
A few nits, and a (I think) a bonafide bug.

File: test_integration/test_plugin_static_policy.py (hunk 4 above)
Comment (bakerbrandond) [reply test_integration/test_plugin_static_policy.py]: inherited from here: https://github.com/geopm/geopm/blob/dev/test_integration/test_static_policy.py#L173-L180
shall we file a bug?

File: test_integration/test_plugin_static_policy.py (hunk 3 above)
Comment (bakerbrandond) [reply test_integration/test_plugin_static_policy.py]: added a decorator to the test class to require the test be ran in an allocation to resolve this (same as the static (non plugin) policy integration test)

File: test_integration/test_plugin_static_policy.py (hunk 4 above)
Comment (bakerbrandond) [reply test_integration/test_plugin_static_policy.py]: There is a subtlety here that we both missed.  The policy name is: POWER_PACKAGE_LIMIT_TOTAL, as in total for all of the packages on the board.  That's why sampling board works but yes this implies they packages split the allotted power equally.  Is a non uniform distribution of power across packages expected to appear in the wild?

File: src/Environment.hpp [hunk 5]
```diff
@@ -107,13 +108,16 @@ namespace geopm
             bool do_profile() const override;
             int timeout(void) const override;
             int debug_attach(void) const override;
-            static std::set<std::string> get_all_vars();
+            static std::set<std::string> get_all_vars(void);
             bool do_ompt(void) const override;
             std::string default_config_path(void) const override;
             std::string override_config_path(void) const override;
+            static void parse_environment_file(const std::string &settings_path,
```
Comment (cmcantalupo) [inline src/Environment.hpp:115]: This static method could be private right?

**Review by cmcantalupo — COMMENTED** [review]
That should work

File: src/Environment.hpp (hunk 5 above)
Comment (bakerbrandond) [reply src/Environment.hpp:115]: Nope the Environment (base class) needs access to this static public member of the EnvironmentImp class line 86 in green of this patch

**Review by bgeltz — COMMENTED** [review]
The integration test looks good now.  Thanks for fixing that up.  I think that bit is g2g.

I assume you and Chris are hashing out the rest, so I didn't approve.

**Pull request merged**